clients/go/gengo.go

```diff
@@ -72,38 +85,196 @@ type WagClient struct {
 	// Keep the circuit doer around so that we can turn it on / off
 	circuitDoer    *circuitBreakerDoer
 	defaultTimeout time.Duration
-	logger       logger.KayveeLogger
+	logger      waglogger.WagClientLogger
 }
 
 var _ Client = (*WagClient)(nil)
 
+
+//This pattern is used instead of using closures for greater transparency and the ability to implement additional interfaces.
+type options struct {
+	transport    http.RoundTripper
+	logger       waglogger.WagClientLogger
+	instrumentor Instrumentor
+	exporter     sdktrace.SpanExporter
+}
+
+type Option interface {
+	apply(*options)
+}
+
+//Logger
+
+//WithLogger sets client logger option.
+func WithLogger(log waglogger.WagClientLogger) Option {
+	return loggerOption{Log: log}
+}
+
+type loggerOption struct {
+	Log waglogger.WagClientLogger
+}
+
+func (l loggerOption) apply(opts *options) {
+	opts.logger = l.Log
+}
+
+//RoundTripper
+
+type roundTripperOption struct {
+	rt http.RoundTripper
+}
+
+func (t roundTripperOption) apply(opts *options) {
+	opts.transport = t.rt
+}
+
+//WithRoundTripper allows you to pass in intrumented/custom roundtrippers which will then wrap the
+//transport roundtripper
+func WithRoundTripper(t http.RoundTripper) Option {
+	return roundTripperOption{rt: t}
+}
+
+//Instrumentor is a function that creates an instrumented round tripper
+type Instrumentor func(baseTransport http.RoundTripper, spanNameCtxValue interface{}, tp sdktrace.TracerProvider) http.RoundTripper
+
+//WithInstrumentor sets a instrumenting function that will be used to wrap the roundTripper for tracing.
+func WithInstrumentor(fn Instrumentor) Option {
+	return instrumentorOption{instrumentor: fn}
+}
+
+type instrumentorOption struct {
+	instrumentor Instrumentor
+}
+
+func (i instrumentorOption) apply(opts *options) {
+	opts.instrumentor = i.instrumentor
+}
+
+//WithExporter sets client span exporter option.
+func WithExporter(se sdktrace.SpanExporter) Option {
+	return exporterOption{exporter: se}
+}
+
+type exporterOption struct {
+	exporter sdktrace.SpanExporter
+}
+
+func (se exporterOption) apply(opts *options) {
+	opts.exporter = se.exporter
+}
+
+//----------------------BEGIN TRACING RELATED FUNCTIONS----------------------
+
+// newResource returns a resource describing this application.
+// Used for setting up tracer provider
+func newResource() *resource.Resource {
+	r, _ := resource.Merge(
+		resource.Default(),
+		resource.NewWithAttributes(
+			semconv.SchemaURL,
+			semconv.ServiceNameKey.String("dapple"),
+			semconv.ServiceVersionKey.String("1.11.0"),
+		),
+	)
+	return r
+}
```

Comment:
Need to remove this as I don't want the semconv dep. Easy to pass {{.ServiceName}} and {{.Version}} to something else (wag/tracing) that then generates this resource. Will make that change today. It's just more passing back and forth that I don't like/can't get away from.